### ISSUE TITLE
Tags to execution requirements propagation in Starlark

### ISSUE THREAD

**Issue by buchgr** [bazelbuild/bazel#7766]
Every target in Bazel supports specifying a `tags` attribute:
```
cc_binary(
  ...
  tags = ["local"]
)
```

While tags can be used for a variety of use cases (i.e. query), Bazel also defines a set of tags that have special meaning for how it executes an action. For example, `no-sandbox` tells Bazel to run this action without a sandboxing strategy and `no-remote` tells it to not run the action on a remote execution system.

The issue is that a user sets `tags` on a target and not on an action. So a rule implementation needs to intentionally pass the tags forward to every action it creates (as an execution requirement).

```
def _impl(ctx):
    ctx.actions.run(
       ...
       execution_requirements = {tag:"" for tag in ctx.attr.tags.keys()},
    )

foo = rule(
    implementation = _impl,
    attrs = {
      ...
    },
)
```

If a rule doesn't add `execution_requirements = ctx.attr.tags.keys()` or similar to each action it creates then a user tagging a target as `no-remote` will have no effect for execution.

We need to answer the following questions:
* Is it desirable for Starlark rules to by default implicitly pass along a defined set of tags (i.e. the ones supported by Bazel) as an execution requirement to every action?
* Would enabling this by default be a backwards incompatible change?
* What are the semantics when a rule also defines their own execution requirements?

cc: @laurentlb @cparsons @nlopezgi @agoulti @jmmv

**Comment by benjaminp** [comment]
An interesting special case today is that `ctx.actions.run_shell` will copy the rule `tags` over into the execution requirements iff you pass a string `command`.

**Comment by ishikhman** [comment]
Design doc [link](https://docs.google.com/document/d/1X2GtuuNT6UqYYOK5lJWQEdPjAgsbdB3nFjjmjso-XHo/edit?ts=5ca70ed0#heading=h.5mcn15i0e1ch)

**Comment by ishikhman** [comment]
was rolled back

**Comment by arlyon** [comment]
@ishikhman is this change applied or is this currently still in progress?

**Comment by ishikhman** [comment]
@arlyon it will be included into our next release, 0.29, which will happen in the beginning of August. 
The change is already in the HEAD though, feel free ti try it out! 
Just add a flag `--incompatible_allow_tags_propagation` to your build.

Note1: flag will be flipped by default in September's release.
Note2: at the moment the change affects only Starlark rules, change for the native rules is coming.
See #8830 for details.
